File: bench/src/logextract/logextract.py
import sys
from collections import Counter
path=sys.argv[1] if len(sys.argv)>1 else "bench/data/access.log"
tl=tb=0
sc={2:0,3:0,4:0,5:0}
hc=[0]*24
ic=Counter()
with open(path,"r",encoding="utf-8",errors="replace") as f:
 for line in f:
  t=line.split(" ")
  if len(t)<10:continue
  ip=t[0]
  hour=int(t[3].split(":")[1])
  status=int(t[8])
  nb=int(t[9])
  tl+=1
  tb+=nb
  sc[status//100]+=1
  hc[hour]+=1
  ic[ip]+=1
o=["=== LOG EXTRACTION RESULTS ===",f"total_lines: {tl}",f"total_bytes: {tb}",f"status_2xx: {sc[2]}",f"status_3xx: {sc[3]}",f"status_4xx: {sc[4]}",f"status_5xx: {sc[5]}"]
for h in range(24):o.append(f"hour_{h:02d}: {hc[h]}")
for i,(ip,c) in enumerate(sorted(ic.items(),key=lambda x:(-x[1],x[0]))[:10],1):o.append(f"top_ip_{i}: {ip} ({c})")
sys.stdout.write("\n".join(o)+"\n")

File: bench/src/logextract/logextract.rs
use std::collections::HashMap;
use std::env;
use std::fs::File;
use std::io::{self,BufRead,BufReader,BufWriter,Write};
fn main(){
 let path=env::args().nth(1).unwrap_or_else(||"bench/data/access.log".to_string());
 let file=File::open(&path).expect("failed to open input file");
 let mut reader=BufReader::new(file);
 let mut tl:u64=0;
 let mut tb:u64=0;
 let mut n2:u64=0;
 let mut n3:u64=0;
 let mut n4:u64=0;
 let mut n5:u64=0;
 let mut hc=[0u64;24];
 let mut ic:HashMap<String,u64>=HashMap::new();
 let mut buf=String::new();
 loop{
  buf.clear();
  if reader.read_line(&mut buf).expect("read error")==0{break;}
  let mut it=buf.split_whitespace();
  let ip=it.next().unwrap_or("");
  it.next();
  it.next();
  let ts=it.next().unwrap_or("");
  it.next();
  it.next();
  it.next();
  it.next();
  let status:u32=it.next().unwrap_or("0").parse().unwrap_or(0);
  let bytes:u64=it.next().unwrap_or("0").parse().unwrap_or(0);
  tl+=1;
  tb+=bytes;
  match status/100{2=>n2+=1,3=>n3+=1,4=>n4+=1,5=>n5+=1,_=>{}}
  let hour=ts.split(':').nth(1).and_then(|h|h.parse::<usize>().ok()).unwrap_or(0);
  if hour<24{hc[hour]+=1;}
  match ic.get_mut(ip){Some(c)=>*c+=1,None=>{ic.insert(ip.to_string(),1);}}
 }
 let mut top:Vec<(String,u64)>=ic.into_iter().collect();
 top.sort_by(|a,b|b.1.cmp(&a.1).then_with(||a.0.cmp(&b.0)));
 let stdout=io::stdout();
 let mut out=BufWriter::new(stdout.lock());
 writeln!(out,"=== LOG EXTRACTION RESULTS ===").unwrap();
 writeln!(out,"total_lines: {}",tl).unwrap();
 writeln!(out,"total_bytes: {}",tb).unwrap();
 writeln!(out,"status_2xx: {}",n2).unwrap();
 writeln!(out,"status_3xx: {}",n3).unwrap();
 writeln!(out,"status_4xx: {}",n4).unwrap();
 writeln!(out,"status_5xx: {}",n5).unwrap();
 for h in 0..24u64{writeln!(out,"hour_{:02}: {}",h,hc[h as usize]).unwrap();}
 for(i,(ip,count))in top.iter().take(10).enumerate(){writeln!(out,"top_ip_{}: {} ({})",i+1,ip,count).unwrap();}
 out.flush().unwrap();
}
